### PR TITLE
replace int by enums in geometry add ring and add part

### PR DESCRIPTION
**Pull request by 3nids**
Shall we use a single enum for all the methods or one enum per type of operation (as done in this PR) ?

Rephrasing is welcome ;)

### CODE REVIEW

**Comment by 3nids** [comment]
@mhugo would be nice to have your feedback since you wrote the TODOs

**Comment by m-kuhn** [comment]
What do you think about having one shared enum for all functions?
I think it would be nice to be able to write the same everywhere

    if not addRing() == QgsGeometry.Success

**Comment by 3nids** [comment]
That's the question I'm asking in this PR ;)

There are pros and cons...
A single enum would make the code clearer, it avoids redundancy but it might get quite long.
Also one drawback is that it's a _hard_ API break since the old integer values won't correspond to the new enum anymore.

Personaly, I don't care. Would be a +0 for single enum.

**Comment by m-kuhn** [comment]
We have been breaking worse than that ;)

**Comment by 3nids** [comment]
@m-kuhn hardest break ever is in place. This seems indeed better this way.

**Comment by wonder-sk** [comment]
If we use a single enum, it would be good to mention in doxygen documentation which are the possible result values. Otherwise the plugin developers either try to handle all errors even if not applicable - or they need to read code to see which ones to handle...

**Comment by 3nids** [comment]
@wonder-sk that's a good point. Would you switch back to several enums?

**Comment by wonder-sk** [comment]
Maybe not necessary to switch to several enums, rather to have one success value shared by all operations and then separate error values for each operation... something like:

enum { Success, AddRingErrorAbc, AddRingErrorDef, AddPartErrorAbc, ... }

**Comment by 3nids** [comment]
OK will do.
Concerning splitPart, using an enum would mean changing up to QgsGeometryEngine (and QgsGeos).
Do you have any concern on this point?

**Comment by wonder-sk** [comment]
Hmm for it I would probably use a separate enum within QgsGeometryEngine and remap them to QgsGeometry error values...

File: python/core/geometry/qgsgeometry.sip [hunk 1]
```diff
@@ -28,8 +28,25 @@ class QgsGeometry
...
+      Success, /*!< Operation succeeded */
+      /* Source geometry or feature issues */
+      NoFeatureSelected, /*!< No feature is selected on the active layer */
+      SeveralFeaturesSelected, /*!< Several features are selected while only one should be */
+      SelectedGeometryNotFound, /*!< The selected geometry cannot be found */
+      NotMultiGeometry, /*!< The source geometry is not multi */
+      /* General input issues */
+      InvalidInputType, /*!< The input geometry (ring or part or ...) has not the correct geometry type */
+      /* Ring specific issues*/
+      RingNotClosed, /*!< The imput ring is not closed */
```
Comment (m-kuhn) [inline python/core/geometry/qgsgeometry.sip]: imput -> input

**Comment by m-kuhn** [comment]
> Maybe not necessary to switch to several enums, rather to have one success value shared by all operations and then separate error values for each operation... something like:
> 
> enum { Success, AddRingErrorAbc, AddRingErrorDef, AddPartErrorAbc, ... }

Not sure about the advantage of this... I'd rather go for either share values wherever possible (and skip the `AddPart`-part) and document methods properly or go for multiple enums.

**Comment by 3nids** [comment]
I added the changes to QgsGeos and QgsGeometryEngine.
I have no clear preference for one or several enums, waiting on other's opinion

**Comment by nyalldawson** [comment]
My vote would go to one enum... Having multiple just seems like a recipe for inconsistent naming

**Comment by m-kuhn** [comment]
Superseded by https://github.com/qgis/QGIS/pull/5014

**Pull request closed without merging**